File: rust/rush-core/src/tracing/flush_worker.rs
//! FlushWorker — background task pool for trace collection and flushing.
//!
//! Mirrors Python's `hush.core.tracing.flush_worker.FlushWorker`.
//! Both collect (CPU-bound) and flush (I/O-bound) run in background
//! tokio tasks, never blocking the caller.

use std::collections::{HashMap, HashSet};
use std::sync::{Arc, Mutex};

use serde_json::Value;
use tokio::task::JoinHandle;

use crate::config::GraphConfig;
use crate::states::state::EngineState;
use crate::tracing::collector::TraceCollector;
use crate::tracing::tracer::Tracer;

/// Background task pool for trace collection and flushing.
///
/// `submit()` spawns a tokio task that collects trace data from state
/// and flushes to all registered tracers. Engine.run_json() calls this
/// and returns immediately.
///
/// Uses interior mutability (Mutex) so `submit()` takes `&self`,
/// allowing the engine to remain `&self` for `run_json()`.
pub struct FlushWorker {
    handles: Mutex<Vec<JoinHandle<()>>>,
}

impl FlushWorker {
    pub fn new() -> Self {
        FlushWorker {
            handles: Mutex::new(Vec::new()),
        }
    }

    /// Submit a collect-and-flush task. Non-blocking — returns immediately.
    pub fn submit(
        &self,
        tracers: Vec<Arc<dyn Tracer>>,
        config: Arc<GraphConfig>,
        state: Arc<EngineState>,
    ) {
        let handle = tokio::spawn(async move {
            // 1. Collect (CPU-bound) — run in blocking thread
            let trace_data = {
                let config = config.clone();
                let state = state.clone();
                match tokio::task::spawn_blocking(move || {
                    let collector = TraceCollector::new(&config);
                    collector.collect(&state, true)
                })
                .await
                {
                    Ok(data) => data,
                    Err(e) => {
                        eprintln!("[rush-tracing] Failed to collect trace data: {e}");
                        return;
                    }
                }
            };

            // 2. Flush to each tracer with merged tags and sampling
            for tracer in &tracers {
                let merged_tags = merge_tags(&trace_data, &tracer.tags());
                let limit = tracer.stream_trace_limit();
                let sampled_nodes = sample_stream_nodes(&trace_data, limit);

                let mut data = trace_data.clone();
                if let Some(obj) = data.as_object_mut() {
                    obj.insert("tags".into(), Value::Array(
                        merged_tags.iter().map(|t| Value::String(t.clone())).collect()
                    ));
                    obj.insert("nodes".into(), sampled_nodes);
                }

                let tracer = tracer.clone();
                // Flush is typically I/O-bound — run in blocking thread
                let _ = tokio::task::spawn_blocking(move || {
                    if let Err(e) = tracer.flush(data) {
                        eprintln!("[rush-tracing] Failed to flush: {e}");
                    }
                })
                .await;
            }
        });
        self.handles.lock().unwrap().push(handle);
    }

    /// Wait for all pending flushes to complete.
    pub async fn wait(&self) {
        let handles: Vec<JoinHandle<()>> = {
            let mut locked = self.handles.lock().unwrap();
            locked.drain(..).collect()
        };
        for handle in handles {
            let _ = handle.await;
        }
    }
}

/// Merge dynamic tags (from state) and static tags (from tracer), deduped.
fn merge_tags(trace_data: &Value, static_tags: &[String]) -> Vec<String> {
    let mut merged: Vec<String> = static_tags.to_vec();
    if let Some(dynamic) = trace_data.get("tags").and_then(|v| v.as_array()) {
        for tag in dynamic {
            if let Some(s) = tag.as_str() {
                if !merged.iter().any(|t| t == s) {
                    merged.push(s.to_string());
                }
            }
        }
    }
    merged
}

/// Apply stream_trace_limit sampling to context groups per generator.
fn sample_stream_nodes(trace_data: &Value, limit: Option<usize>) -> Value {
    let nodes = match trace_data.get("nodes").and_then(|v| v.as_array()) {
        Some(arr) => arr,
        None => return Value::Array(vec![]),
    };

    let limit = match limit {
        Some(l) => l,
        None => return Value::Array(nodes.clone()),
    };

    // Find parents that contain generators
    let mut gen_parents: HashSet<String> = HashSet::new();
    for n in nodes {
        if n.get("kind").and_then(|v| v.as_str()) == Some("generator") {
            if let Some(parent) = n.get("parent_trace_key").and_then(|v| v.as_str()) {
                gen_parents.insert(parent.to_string());
            }
        }
    }

    if gen_parents.is_empty() {
        return Value::Array(nodes.clone());
    }

    // Count top-level stream_context groups per generator parent
    let mut counts: HashMap<String, usize> = HashMap::new();
    let mut remove_keys: HashSet<String> = HashSet::new();

    for n in nodes {
        if n.get("kind").and_then(|v| v.as_str()) == Some("stream_context") {
            if let Some(parent) = n.get("parent_trace_key").and_then(|v| v.as_str()) {
                if gen_parents.contains(parent) {
                    let count = counts.entry(parent.to_string()).or_insert(0);
                    *count += 1;
                    if *count > limit {
                        if let Some(key) = n.get("trace_key").and_then(|v| v.as_str()) {
                            remove_keys.insert(key.to_string());
                        }
                    }
                }
            }
        }
    }

    if remove_keys.is_empty() {
        return Value::Array(nodes.clone());
    }

    // Cascade: remove all descendants of removed context groups
    let mut children_map: HashMap<String, Vec<String>> = HashMap::new();
    for n in nodes {
        if let (Some(parent), Some(key)) = (
            n.get("parent_trace_key").and_then(|v| v.as_str()),
            n.get("trace_key").and_then(|v| v.as_str()),
        ) {
            children_map
                .entry(parent.to_string())
                .or_default()
                .push(key.to_string());
        }
    }

    let mut queue: Vec<String> = remove_keys.iter().cloned().collect();
    while let Some(k) = queue.pop() {
        if let Some(children) = children_map.get(&k) {
            for child in children {
                if remove_keys.insert(child.clone()) {
                    queue.push(child.clone());
                }
            }
        }
    }

    let filtered: Vec<Value> = nodes
        .iter()
        .filter(|n| {
            n.get("trace_key")
                .and_then(|v| v.as_str())
                .map_or(true, |k| !remove_keys.contains(k))
        })
        .cloned()
        .collect();

    Value::Array(filtered)
}

File: operon/core/tracing/flush_worker.py
"""FlushWorker — background thread pool for trace collection and flushing.

Both collect (CPU-bound) and flush (I/O-bound) run in the thread pool,
never blocking the main async thread.
"""

import atexit
import logging
from collections import defaultdict
from concurrent.futures import Future, ThreadPoolExecutor
from typing import Any, Dict, List, Optional

from operon.core.tracing.base import Tracer
from operon.core.tracing.collector import TraceCollector

LOGGER = logging.getLogger("operon.tracing")

_worker: "FlushWorker | None" = None


class FlushWorker:
    """Thread pool for trace collection and flushing.

    Runs both TraceCollector.collect() and tracer.flush() in background threads
    so engine.run() returns immediately without blocking.
    """

    def __init__(self, max_workers: int = 4):
        self._executor = ThreadPoolExecutor(
            max_workers=max_workers, thread_name_prefix="operon-trace"
        )
        self._futures: List[Future] = []

    def submit(self, tracers: List[Tracer], collector: TraceCollector, state: Any) -> Future:
        """Submit a collect-and-flush task to the thread pool.

        Returns a Future that callers can optionally wait on to check for errors.
        The background thread will:
        1. Run collector.collect(state)
        2. For each tracer, merge tags and call tracer.flush()

        Args:
            tracers: List of Tracer instances to flush to
            collector: Pre-built TraceCollector (graph metadata already computed)
            state: MemoryState after execution completes

        Returns:
            Future that resolves when flush completes. Call .result() to
            re-raise any exception from the background thread.
        """
        self._futures = [f for f in self._futures if not f.done()]
        future = self._executor.submit(self._collect_and_flush, tracers, collector, state)
        self._futures.append(future)
        return future

    def _collect_and_flush(
        self, tracers: List[Tracer], collector: TraceCollector, state: Any
    ) -> None:
        """Collect trace data and flush to all tracers.

        Errors are logged AND re-raised so they propagate through the Future.
        """
        errors: List[Exception] = []

        try:
            # 1. Collect trace tree (CPU-bound, microseconds)
            trace_data = collector.collect(state)
        except Exception as e:
            LOGGER.exception("Failed to collect trace data")
            raise

        # 2. Flush to each tracer with merged tags
        for tracer in tracers:
            try:
                # Merge: dynamic tags (from state) + static tags (from tracer)
                merged = _merge_tags(trace_data.get("tags", []), tracer.tags)
                # Apply stream sampling per tracer's limit
                nodes = trace_data.get("nodes", [])
                limit = getattr(tracer, "_stream_trace_limit", None)
                sampled_nodes = _sample_stream_nodes(nodes, limit)
                # Apply trace filter (exclude noisy ops, skip empty, etc.)
                tf = getattr(tracer, "trace_filter", None)
                filtered_nodes = tf.apply(sampled_nodes) if tf else sampled_nodes
                # Create a copy with merged tags and filtered nodes
                data = {**trace_data, "tags": merged if merged else None, "nodes": filtered_nodes}
                tracer.flush(data)
            except Exception as e:
                LOGGER.exception("Failed to flush traces to %s", type(tracer).__name__)
                errors.append(e)

        if errors:
            msg = f"Flush failed for {len(errors)} tracer(s): {errors}"
            raise RuntimeError(msg) from errors[0]

    def wait(self, timeout: Optional[float] = None) -> List[Exception]:
        """Wait for all pending flush tasks to complete.

        Args:
            timeout: Max seconds to wait per future. None = wait forever.

        Returns:
            List of exceptions from failed flushes (empty if all succeeded).
        """
        errors = []
        for future in self._futures:
            try:
                future.result(timeout=timeout)
            except Exception as e:
                errors.append(e)
        self._futures.clear()
        return errors

    def shutdown(self, wait: bool = True) -> None:
        """Shutdown the thread pool."""
        self._executor.shutdown(wait=wait)


def _sample_stream_nodes(nodes: List[Dict[str, Any]], limit: Optional[int]) -> List[Dict[str, Any]]:
    """Apply stream_trace_limit sampling to context groups per generator.

    For each graph that contains generators, caps the number of top-level
    stream_context groups. Excess context groups and all their descendants
    are removed.

    Args:
        nodes: Full list of TraceNode dicts (from collect_tree).
        limit: Max context groups to keep per generator parent.
            None = keep all, 0 = drop all context groups.

    Returns:
        Filtered node list. Generator and non-stream nodes pass through
        unchanged.
    """
    if limit is None:
        return nodes

    # Find parents that contain generators
    gen_parents: set = set()
    for n in nodes:
        if n.get("kind") == "generator":
            parent = n.get("parent_trace_key")
            if parent:
                gen_parents.add(parent)

    if not gen_parents:
        return nodes

    # Count top-level stream_context groups per generator parent
    counts: Dict[str, int] = defaultdict(int)
    remove_keys: set = set()

    for n in nodes:
        if n.get("kind") == "stream_context" and n.get("parent_trace_key") in gen_parents:
            parent = n["parent_trace_key"]
            counts[parent] += 1
            if counts[parent] > limit:
                remove_keys.add(n["trace_key"])

    if not remove_keys:
        return nodes

    # Cascade: remove all descendants of removed context groups
    children_map: Dict[str, list] = defaultdict(list)
    for n in nodes:
        p = n.get("parent_trace_key")
        if p:
            children_map[p].append(n["trace_key"])

    queue = list(remove_keys)
    while queue:
        k = queue.pop(0)
        for child in children_map.get(k, []):
            remove_keys.add(child)
            queue.append(child)

    return [n for n in nodes if n["trace_key"] not in remove_keys]


def _merge_tags(dynamic_tags: List[str], static_tags: List[str]) -> List[str]:
    """Merge dynamic (from state) and static (from tracer) tags, deduped.

    Static tags come first, then unique dynamic tags are appended.
    Same logic as existing BaseTracer._merge_tags().
    """
    merged = list(static_tags)
    for tag in dynamic_tags:
        if tag not in merged:
            merged.append(tag)
    return merged


def get_flush_worker() -> FlushWorker:
    """Get or create the global FlushWorker singleton.

    The worker is created lazily on first call and registered for
    shutdown at interpreter exit.
    """
    global _worker
    if _worker is None:
        _worker = FlushWorker()
        atexit.register(_worker.shutdown)
    return _worker
